src/http/request.cr

```diff
@@ -121,6 +117,89 @@ class HTTP::Request
     BadRequest.new
   end
 
+  private METHODS = %w(GET HEAD POST PUT DELETE CONNECT OPTIONS PATCH TRACE)
+
+  private def self.parse_request_line(io : IO) : RequestLine | BadRequest | Nil
+    # Optimization: see if we have a peek buffer
+    # (avoids a string allocation for the entire request line)
+    if peek = io.peek
+      # peek.empty? means there's no more input (EOF), so no more requests
+      return nil if peek.empty?
+
+      # See if we can find \r\n (first \n, then \r before it)
+      index = peek.index('\n'.ord.to_u8)
+      # TODO: eventually increase the 4096 bytes limit
+      if index && index > 0 && index < 4096 && peek[index - 1] == '\r'.ord.to_u8
+        parts = parse_request_line(peek[0, index - 1])
+        io.skip(index + 1) # Must skip until after \r\n
+        return parts
+      end
+    end
+
+    request_line = io.gets(4096, chomp: true)
+    return BadRequest.new unless request_line
+
+    parse_request_line(request_line)
+  end
+
+  private def self.parse_request_line(line : String) : RequestLine | BadRequest
+    parse_request_line(line.to_slice)
+  end
+
+  private def self.parse_request_line(slice : Bytes) : RequestLine | BadRequest
+    space_index = slice.index(' '.ord.to_u8)
+
+    # Oops, only a single part (should be three)
+    return BadRequest.new unless space_index
+
+    subslice = slice[0...space_index]
+
+    # Optimization: see if it's one of the common methods
```

Comment:
Is using the `StringPool` hashing method faster or slower than searching the array here? I suspect this is faster, but I could be wrong.

Reply:
Of course, the fastest way is probably a manually constructed parse tree like the crystal lexer uses. That can be added later.

Reply:
I don't think we need a `StringPool` for this because the number of methods is very small. A parse tree is also an option, if someone else wants to try it they can do it but I think what takes most of the time is the string allocation and the IO, not the actual parsing of the method.

Reply:
Plus a `StringPool` would means people could keep sending different methods, eventually consuming all memory because of it.